### PR TITLE
Port MergeRelationshipAcceptanceTest to TCK

### PR DESCRIPTION
**Pull request by Mats-SX**
Add new error detail: `VariableAlreadyBound`.

### CODE REVIEW

File: tck/features/MergeRelationshipAcceptance.feature [hunk 1]
```diff
@@ -0,0 +1,833 @@
...
+    And having executed:
+      """
+      CREATE (:S)
+      CREATE (:End {prop: 42})
+      """
+    When executing query:
+      """
+      MATCH (a:S)
+      FOREACH(x IN [42] |
+        MERGE (b:End {prop: x})
```
Comment (petraselmer) [inline tck/features/MergeRelationshipAcceptance.feature]: This (indenting after `FOREACH`) is nice... should this be added to the style guide too?

File: tck/features/MergeRelationshipAcceptance.feature [hunk 2]
```diff
@@ -0,0 +1,833 @@
...
+      MATCH (a:A), (b:B)
+      MERGE (a)-[r:TYPE {name: 'r2'}]->(b)
+      RETURN count(r)
+      """
+    Then the result should be:
+      | count(r) |
+      | 1        |
+    And no side effects
+
+  Scenario: Creating relationship when all matches filtered out
```
Comment (petraselmer) [inline tck/features/MergeRelationshipAcceptance.feature]: Is there a better name for this scenario? It is not very clear at the moment...

File: tck/features/MergeRelationshipAcceptance.feature [hunk 3]
```diff
@@ -0,0 +1,833 @@
...
+      RETURN count(r)
+      """
+    Then the result should be:
+      | count(r) |
+      | 1        |
+    And the side effects should be:
+      | +relationships | 1 |
+      | +properties    | 1 |
+
+  Scenario: Using `ON CREATE`, node
```
Comment (petraselmer) [inline tck/features/MergeRelationshipAcceptance.feature]: Can this scenario name be improved?

File: tck/features/MergeRelationshipAcceptance.feature [hunk 4]
```diff
@@ -0,0 +1,833 @@
...
+      MERGE (a)-[:KNOWS]->(b)
+        ON CREATE SET b.created = timestamp()
+      """
+    Then the result should be empty
+    And the side effects should be:
+      | +nodes         | 2 |
+      | +relationships | 1 |
+      | +properties    | 1 |
+
+  Scenario: Using `ON CREATE`, relationship
```
Comment (petraselmer) [inline tck/features/MergeRelationshipAcceptance.feature]: Can this scenario name be improved?

File: tck/features/MergeRelationshipAcceptance.feature [hunk 5]
```diff
@@ -0,0 +1,833 @@
...
+      """
+      MERGE (a)-[r:KNOWS]->(b)
+        ON MATCH SET r.created = timestamp()
+      """
+    Then the result should be empty
+    And the side effects should be:
+      | +nodes         | 2 |
+      | +relationships | 1 |
+
+  Scenario: Using `ON MATCH`, relationship
```
Comment (petraselmer) [inline tck/features/MergeRelationshipAcceptance.feature]: Can this scenario name be made clearer?

File: tck/features/MergeRelationshipAcceptance.feature [hunk 6]
```diff
@@ -0,0 +1,833 @@
...
+      CREATE (:Person {id: 666})
+      """
+    When executing query:
+      """
+      CREATE (a:A)
+      MERGE (a)-[:KNOWS]->(b:Person {id:666})
+      """
+    Then a ConstraintValidationFailed should be raised at runtime: CreateBlockedByConstraint
+
+  Scenario: Inside `FOREACH`, after match
```
Comment (petraselmer) [inline tck/features/MergeRelationshipAcceptance.feature]: Can this scenario name be improved?

File: tck/features/MergeRelationshipAcceptance.feature [hunk 7]
```diff
@@ -0,0 +1,833 @@
...
+        MERGE (a)-[:FOO]->({prop: x})
+      )
+      """
+    Then the result should be empty
+    And the side effects should be:
+      | +nodes         | 2 |
+      | +relationships | 2 |
+      | +properties    | 2 |
+
+  Scenario: Inside `FOREACH`, after match, following other merge
```
Comment (petraselmer) [inline tck/features/MergeRelationshipAcceptance.feature]: Can this scenario name be improved?

File: tck/features/MergeRelationshipAcceptance.feature [hunk 8]
```diff
@@ -0,0 +1,833 @@
...
+      FOREACH(x IN [42] |
+        MERGE (b:End {prop: x})
+        MERGE (a)-[:FOO]->(b)
+      )
+      """
+    Then the result should be empty
+    And the side effects should be:
+      | +relationships | 1 |
+
+  Scenario: Inside `FOREACH`, no match, following other merges
```
Comment (petraselmer) [inline tck/features/MergeRelationshipAcceptance.feature]: Can this scenario name be improved?

File: tck/features/MergeRelationshipAcceptance.feature [hunk 9]
```diff
@@ -0,0 +1,833 @@
...
+        MERGE (a)-[:FOO]->(b)
+      )
+      """
+    Then the result should be empty
+    And the side effects should be:
+      | +nodes         | 3 |
+      | +relationships | 2 |
+      | +properties    | 3 |
+
+  Scenario: Inside `FOREACH`, after with, following other merges
```
Comment (petraselmer) [inline tck/features/MergeRelationshipAcceptance.feature]: Can this scenario name be improved?

File: tck/features/MergeRelationshipAcceptance.feature [hunk 10]
```diff
@@ -0,0 +1,833 @@
...
+      """
+      MERGE ({name: 'Andres'})-[:R]->({name: 'Emil'})
+      """
+    Then the result should be empty
+    And the side effects should be:
+      | +nodes         | 2 |
+      | +relationships | 1 |
+      | +properties    | 2 |
+
+  Scenario: Inside nested foreach
```
Comment (petraselmer) [inline tck/features/MergeRelationshipAcceptance.feature]: Capitalized `FOREACH`?

File: tck/features/MergeRelationshipAcceptance.feature [hunk 11]
```diff
@@ -0,0 +1,833 @@
...
+        )
+      )
+      """
+    Then the result should be empty
+    And the side effects should be:
+      | +nodes         | 16 |
+      | +relationships | 24 |
+      | +properties    | 32 |
+
+  Scenario: Inside nested foreach, nodes inlined
```
Comment (petraselmer) [inline tck/features/MergeRelationshipAcceptance.feature]: Capitalized `FOREACH`?

File: tck/features/MergeRelationshipAcceptance.feature [hunk 12]
```diff
@@ -0,0 +1,833 @@
...
+    Given an empty graph
+    When executing query:
+      """
+      FOREACH(x IN [0, 1, 2] |
+        FOREACH(y IN [0, 1, 2] |
+          MERGE (a {x: x, y: y})-[:R]->(b {x: x + 1, y: y})
+          MERGE (c {x: x, y: y + 1})-[:R]->(d {x: x + 1, y: y + 1})
+          MERGE (a)-[:R]->(c) MERGE (b)-[:R]->(d)
```
Comment (petraselmer) [inline tck/features/MergeRelationshipAcceptance.feature]: The second `MERGE` clause ought to be on a new line

File: tck/features/MergeRelationshipAcceptance.feature [hunk 13]
```diff
@@ -0,0 +1,833 @@
...
+      """
+    Then a SyntaxError should be raised at compile time: VariableAlreadyBound
+
+  Scenario: Using list properties via variable
+    Given an empty graph
+    When executing query:
+      """
+      CREATE (a:Foo), (b:Bar)
+      WITH a, b
+      UNWIND ["a,b", "a,b"] AS str
```
Comment (petraselmer) [inline tck/features/MergeRelationshipAcceptance.feature]: -> `UNWIND ['a,b', 'a,b'] AS str`

File: tck/features/MergeRelationshipAcceptance.feature [hunk 14]
```diff
@@ -0,0 +1,833 @@
...
+      WITH a, b, split(str, ",") AS roles
```
Comment (petraselmer) [inline tck/features/MergeRelationshipAcceptance.feature]: -> WITH a, b, split(str, ',') AS roles

File: tck/features/MergeRelationshipAcceptance.feature [hunk 15]
```diff
@@ -0,0 +1,833 @@
...
+      RETURN count(a)
+      """
+    Then the result should be:
+      | count(a) |
+      | 1        |
+    And the side effects should be:
+      | +nodes         | 2 |
+      | +relationships | 1 |
+
+  Scenario: `UNWIND` with multiple merges
```
Comment (petraselmer) [inline tck/features/MergeRelationshipAcceptance.feature]: Are the backticks needed in a scenario name? If yes, we should standardize these...if not, I think we ought to remove them

File: tck/features/MergeRelationshipAcceptance.feature [hunk 16]
```diff
@@ -0,0 +1,833 @@
...
+    Given an empty graph
+    When executing query:
+      """
+      UNWIND ['Keanu Reeves', 'Hugo Weaving', 'Carrie-Anne Moss', 'Laurence Fishburne'] AS actor
+      MERGE (m:Movie  {name: 'The Matrix'})
```
Comment (petraselmer) [inline tck/features/MergeRelationshipAcceptance.feature]: I think there may be a double space between "m:Movie" and "{name:...}"

File: tck/features/MergeRelationshipAcceptance.feature (hunk 2 above)
Comment (Mats-SX) [inline tck/features/MergeRelationshipAcceptance.feature]: I think this name is good.

File: tck/features/MergeRelationshipAcceptance.feature (hunk 15 above)
Comment (Mats-SX) [inline tck/features/MergeRelationshipAcceptance.feature]: They are not needed. I can't recall why I started using them.

They are necessary when using parentheses, for things like functions.

File: tck/features/MergeRelationshipAcceptance.feature (hunk 16 above)
Comment (Mats-SX) [inline tck/features/MergeRelationshipAcceptance.feature]: Well spotted!

File: tck/features/MergeRelationshipAcceptance.feature (hunk 1 above)
Comment (Mats-SX) [inline tck/features/MergeRelationshipAcceptance.feature]: Yes, let's propose it!

**Pull request merged**